### PR TITLE
Add storage API to open a file for appending, without truncating it's content first

### PR DESCRIPTION
**Pull request by mbaklor**
Implemented in both memory and desktop, need to implement for mobile still, setting this as a draft for now, I don't know who can do mobile and how but for now this is what I could pull off. 

 

### Description:

Needed an option to add to a file in the storage without truncating it first, specifically to write long lasting log files.

Fixes #(issue)

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

#### Where applicable:

- [x] Public APIs match existing style and have Since: line.

### CODE REVIEW

**Review by andydotxyz — COMMENTED** [review]
Clear and clean, thanks for looking at this - a good addition.

I think it will need the mobile code added to land though.
Would tips on how to progress be sufficient, or are you handing this on to someone else?

**Comment by mbaklor** [comment]
Thank you that was super encouraging and got me the drive to try the mobile code myself! 

Let me know if this mobile code makes sense to you, and if someone with an android or ios setup - well - set up could check that this code works I'd super appreciate it as well!

**Comment by andydotxyz** [comment]
Looking great. The code compiles and runs on my iOS test devices.
But there is no test for the file repository on mobile.
Is there a snippet of code that could be run, or an example we could drop into fyne_demo that would verify?

Thanks!

**Comment by mbaklor** [comment]
I have discovered in writing this test snippet that I in fact forgot to implement the last part of the puzzle, the storage interface itself. Fixed and pushed, and here's the snippet I used to check it on my PC

```go
func main() {
	app := app.NewWithID("io.fyne.appendtest")
	win := app.NewWindow("Append Test")
	win.Resize(fyne.NewSize(400, 300))

	app.Storage().Create("test.txt")

	btnAppend := widget.NewButton("append to file", func() {
		doc, err := app.Storage().Append("test.txt")
		if err != nil {
			dialog.ShowError(err, win)
			return
		}
		_, err = doc.Write([]byte("test line\n"))
		if err != nil {
			dialog.ShowError(err, win)
			return
		}
		err = doc.Close()
		if err != nil {
			dialog.ShowError(err, win)
			return
		}
		dialog.ShowInformation("Success", "Wrote to file", win)
	})
	btnRead := widget.NewButton("read file", func() {
		doc, err := app.Storage().Open("test.txt")
		if err != nil {
			dialog.ShowError(err, win)
			return
		}
		b := make([]byte, 64)
		_, err = doc.Read(b)
		if err != nil {
			dialog.ShowError(err, win)
			return
		}
		dialog.ShowInformation("Success", "Successfully read from file\n\ncontents:\n"+string(b), win)
		println(string(b))
	})

	content := container.NewGridWithRows(2, btnAppend, btnRead)
	win.SetContent(container.NewCenter(content))
	win.ShowAndRun()
}
```

File: storage.go [hunk 1]
```diff
@@ -8,6 +8,7 @@ type Storage interface {
 	Create(name string) (URIWriteCloser, error)
 	Open(name string) (URIReadCloser, error)
 	Save(name string) (URIWriteCloser, error)
+	Append(name string) (URIWriteCloser, error)
```
Comment (andydotxyz) [inline storage.go:11]: Making this change into an existing interface is a breaking change and we cannot do it (unless a v3 is made).
Additional features added on top of the initial storage API will have to be optional.

**Review by andydotxyz — CHANGES_REQUESTED** [review]

File: storage.go (hunk 1 above)
Comment (mbaklor) [reply storage.go:11]: Okay, fair, didn't think about how people who implemented the interface themselves will lose it like this. In that case though, how do we make this addition optional while allowing for `app.Storage().Append` or other convenient usage of the new API?

Comment (dweymouth) [reply storage.go:11]: You can create a new interface like Appender or StorageAppender that defines the single function. And then the builtin one implements both Storage and Appender

Comment (mbaklor) [reply storage.go]: for now I deleted this commit, as there's no real reason to change the storage interface at all, I'll think about how to deal with the `internal.Docs` questions that came up for me here but for now I can resolve this at least

**Comment by mbaklor** [comment]
the new "it works on my system" with the storage interface reverted to how it was is:

```go
func main() {
	app := app.NewWithID("io.fyne.appendtest")
	win := app.NewWindow("Append Test")
	win.Resize(fyne.NewSize(400, 300))
	root := app.Storage().RootURI()
	uri, err := storage.Child(root, "test.txt")
	if err != nil {
		dialog.ShowError(err, win)
		return
	}

	btnAppend := widget.NewButton("append to file", func() {
		doc, err := storage.Appender(uri)
		defer doc.Close()
		if err != nil {
			dialog.ShowError(err, win)
			return
		}
		_, err = doc.Write([]byte("test line\n"))
		if err != nil {
			dialog.ShowError(err, win)
			return
		}
		dialog.ShowInformation("Success", "Wrote to file", win)
	})
	btnRead := widget.NewButton("read file", func() {
		doc, err := storage.Reader(uri)
		defer doc.Close()
		if err != nil {
			dialog.ShowError(err, win)
			return
		}
		b := make([]byte, 64)
		_, err = doc.Read(b)
		if err != nil {
			dialog.ShowError(err, win)
			return
		}
		dialog.ShowInformation("Success", "Successfully read from file\n\ncontents:\n"+string(b), win)
		println(string(b))
	})

	content := container.NewGridWithRows(2, btnAppend, btnRead)
	win.SetContent(container.NewCenter(content))
	win.ShowAndRun()
}
```

File: internal/driver/mobile/repository.go [hunk 2]
```diff
@@ -67,5 +67,11 @@ func (m *mobileFileRepo) Reader(u fyne.URI) (fyne.URIReadCloser, error) {
 }
 
 func (m *mobileFileRepo) Writer(u fyne.URI) (fyne.URIWriteCloser, error) {
-	return fileWriterForURI(u)
+	return fileWriterForURI(u, true)
+}
+
+// TODO: need someone who understands native code to write this, I'm not sure
```
Comment (andydotxyz) [inline internal/driver/mobile/repository.go:73]: Seems like this TODO could go?

Comment (mbaklor) [reply internal/driver/mobile/repository.go:73]: OOPS! deleted

**Comment by andydotxyz** [comment]
This doesn't work work on Android - I finally tracked it down to the `modes` always being `NULL` which caused all write operations to fail.

This patch fixes Android!

```diff
diff --git a/internal/driver/mobile/android.c b/internal/driver/mobile/android.c
index 14e72b66b..db4f55c06 100644
--- a/internal/driver/mobile/android.c
+++ b/internal/driver/mobile/android.c
@@ -154,9 +154,9 @@ void* saveStream(uintptr_t jni_env, uintptr_t ctx, char* uriCstr, bool truncate)
        jobject uri = parseURI(jni_env, ctx, uriCstr);
        jstring modes = NULL;
        if (truncate) {
-               jstring modes = (*env)->NewStringUTF(env, "wt"); // truncate before write
+               modes = (*env)->NewStringUTF(env, "wt"); // truncate before write
        } else {
-               jstring modes = (*env)->NewStringUTF(env, "wa");
+               modes = (*env)->NewStringUTF(env, "wa");
        }
        jobject stream = (jobject)(*env)->CallObjectMethod(env, resolver, saveOutputStream, uri, modes);
        jthrowable loadErr = (*env)->ExceptionOccurred(env);
```

**Comment by andydotxyz** [comment]
Unfortunately on iOS it reports writing successfully but the read back just gets EOF not the file content

**Comment by andydotxyz** [comment]
The following change made it work on iOS:

```diff
diff --git a/internal/driver/mobile/file_ios.m b/internal/driver/mobile/file_ios.m
index 18b08de24..9ff83770d 100644
--- a/internal/driver/mobile/file_ios.m
+++ b/internal/driver/mobile/file_ios.m
@@ -39,7 +39,7 @@ bool iosExistsPath(const char* path) {
     NSFileHandle* handle = [NSFileHandle fileHandleForWritingToURL:url error:&err];
 
     if (!truncate) {
-        [handle truncateFileAtOffset:[handle seekToEndOfFile]];
+        [handle seekToEndOfFile];
     }
 
     return handle;
```

but in part it was because the read operation can return EOF error which we were treating as a problem, so try this:

```go
              _, err = doc.Read(b)
              if err != nil && err != io.EOF {
                      dialog.ShowError(err, win)
                      return
              }
```

**Review by andydotxyz — APPROVED** [review]
Thanks so much for this

**Pull request merged**